src/state.rs

```diff
@@ -456,6 +526,54 @@ impl DemandProxyState {
             .cloned()
     }
 
+    pub fn set_cached_resolve_dns_for_hostname(
+        &mut self,
+        hostname: String,
+        cached_atom_bool: Arc<SingleFlight>,
+    ) {
+        self.state
+            .write()
+            .unwrap()
+            .cached_resolved_dns_map
+            .crd_hashmap
+            .insert(hostname, cached_atom_bool);
+    }
+
+    pub fn get_cached_resolve_dns_for_hostname(
+        &mut self,
+        hostname: &String,
+    ) -> Option<Arc<SingleFlight>> {
+        self.state
+            .read()
+            .unwrap()
+            .cached_resolved_dns_map
+            .crd_hashmap
+            .get(hostname)
+            .cloned()
+    }
+
+    // loop_for_resolve_dns help to implement a duplicate function call suppression mechanism.
+    pub fn loop_for_resolve_dns(&self, cached_atom_bool: Arc<SingleFlight>) {
+        loop {
+            // Set the timeout duration for each same hostname request
+            let timeout_duration = Duration::from_secs(10);
+            // Start the timer
+            let start_time = Instant::now();
+            if !cached_atom_bool
+                .complete_resolved_dns
+                .load(Ordering::Relaxed)
+            {
+                let elapsed_time = start_time.elapsed();
+                // Check if the timer is timed out or not
+                if elapsed_time >= timeout_duration {
```

Comment:
dead loop until >10s or has complete_resolved_dns is true

Reply:
Hi @hzxuzhonghu, yes, this dead loop will break in 2 cases:

1. timeout after 10s
2. complete_resolved_dns is true

I’m not sure that 10 seconds is appropriate or not, any suggestion?

Reply:
The critical point is cpu wasting polling on a value, here it should be waiting for a signal/sem

Reply:
Hi @hzxuzhonghu, get your points now. I change the code by using `use tokio::sync::Notify;` to implement the notifications for other requests. thanks for your advice!
I also have some details to discuss with you @howardjohn & @hzxuzhonghu, it may be not suitable to use MPSC channel in this use case after some experiments. As you know, MPSC stands for "Multiple Producers and Single Consumer", but in this case it's more like "Single Producers and Multiple Consumer", so `tokio::sync::Notify` should be the more proper. Thanks!